torch/optim/lr_scheduler.py

```diff
@@ -582,18 +647,40 @@ def __init__(
         super().__init__(optimizer, last_epoch)
 
     @override
-    def get_lr(self) -> list[float]:
-        """Compute the learning rate of each parameter group."""
+    def get_lr(self) -> list[float | Tensor]:
+        """Compute the next learning rate for each of the optimizer's :attr:`~torch.optim.Optimizer.param_groups`.
+
+        If the current epoch is in :attr:`milestones`, decays the ``group["lr"]``s in the optimizer's
+        :attr:`~torch.optim.Optimizer.param_groups` by :attr:`gamma`.
+
+        Returns:
+            list[float | Tensor]: A :class:`list` of learning rates for each of the optimizer's
+            :attr:`~torch.optim.Optimizer.param_groups` with the same types as their current ``group["lr"]``s.
+
+        .. note:: If you're trying to inspect the most recent learning rate, use :meth:`get_last_lr()` instead.
+
+        .. note:: If the current epoch appears in :attr:`milestones` ``n`` times, we scale by :attr:`gamma` to the power
+        of ``n``
+        """
         _warn_get_lr_called_within_step(self)
 
         if self.last_epoch not in self.milestones:
-            return [group["lr"] for group in self.optimizer.param_groups]
+            return _param_groups_val_list(self.optimizer, "lr")
         return [
             group["lr"] * self.gamma ** self.milestones[self.last_epoch]
             for group in self.optimizer.param_groups
         ]
```

Comment:
For closed-factor stuff, I propose to factor out these to e.g. `torch.optim.scheduler._step_scheduler(...)` - so that they can in the future be used directly without having to create `LRScheduler` first. (and existing `_get_closed_form_lr` could also then use `torch.optim.scheduler._step_scheduler(...)` internally)

Reply:
I'm not sure I understand. Intuitively, wouldn't we want to keep most computation in some functional `def compute_lr(step, total_iters, base_lr, ...)` and turn the "stateful per-iteration stepping" into a thin wrapper around that? Meaning we wouldn't need these incremental forms anymore.

Maybe we can discuss further in a separate issue - #125531, #67760, or a new one.